doc/phar-signing.md

```diff
@@ -0,0 +1,364 @@
+other binary data. It is analogous to the [JAR files][jar], and will require PHP to execute the PHARs. The content
+of a PHAR is as follows:
+
+1. A stub: a piece of code that handles the extraction of resources.
+1. A binary manifest which allows the interpreter to understand the file structure of the embedded contents that follow.
+1. The actual content of the archive.
+1. The signature makes up the last section of the file. It is a 4-byte signature flag which tells what signature type
+   was used and then another 4-byte constant marks the file as having a signature.
+
+When PHP later reads the archive, it can determine the signature and type by reading the end of the archive. This way,
+if the content of the PHAR has been tempered with, e.g. code was injected in the archive, PHP will see that the content
+of the archive does not match the signature of the PHAR and will bail out.
+
+
+### Why it is bad
+
+If a PHAR gets corrupted, the PHP interpreter will know from the signature and throw a `PharException`.
+
+The first issue is the signature itself: if the PHAR gets corrupted, maybe the signature got corrupted too. Using an
+OpenSSL based signature will help to prevent this issue however.
+
+The second issue is the mechanism itself. Indeed, if one injects code _before_ the stub, then this code will be executed
+before the signature check.
+
+**This security mechanism cannot prevent modifications of the archive itself.**  
+
+The good news, there is a solution.
+
+
+## How to (properly) sign your PHAR
+
+### Create a new GPG-key
+
+The first step is to create a new GPG-key. You can either do that via a GUI or via the CLI like this:
+
+```
+$ gpg --gen-key
+```
+
+It will ask for some questions. It is recommended to use a passphrase (ideally generated and managed by a reputable
+password manager). In the end, you will end up with something like this:
+
+```
+pub   ed25519 2023-10-21 [SC] [expires: 2026-10-20]
+      96C8013A3CC293C465EE3FBB03B2F4DF7A20DF08
+uid                      Théo Fidry <theo.fidry+phar-signing-example@example.com>
+sub   cv25519 2023-10-21 [E] [expires: 2026-10-20]
+```
+
+In this case the interesting part is `96C8013A3CC293C465EE3FBB03B2F4DF7A20DF08` which is the key ID. You can also check
+the list of your GPG keys like so:
+
+```
+$ gpg --list-secret-keys --keyid-format=long
+
+#
+# Other keys displayed too
+#
+sec   ed25519/03B2F4DF7A20DF08 2023-10-21 [SC] [expires: 2026-10-20]
+      96C8013A3CC293C465EE3FBB03B2F4DF7A20DF08
+uid                 [ultimate] Théo Fidry <theo.fidry+phar-signing-example@example.com>
+ssb   cv25519/765C0E3CCBC7D7D3 2023-10-21 [E] [expires: 2026-10-20]
+```
+
+The interesting part is the `96C8013A3CC293C465EE3FBB03B2F4DF7A20DF08` which is the key-ID.
+
+To make the key accessible for others we should now send it to a keyserver.
+
+```
+$ gpg --keyserver pgp.mit.edu --send-key 96C8013A3CC293C465EE3FBB03B2F4DF7A20DF08
+```
+
```

Comment:
Please push things to [keys.openpgp.org](https://keys.openpgp.org/). They will ensure actual ownership of the email-address used and thus add a very significant layer of additional security before making the pub key available.

If you want to, you can, for legacy reasons, push things to pgp.mit.edu and friends additionally.